src/libraries/Common/src/System/Net/Http/aspnetcore/Http2/Hpack/Huffman.cs

```diff
@@ -307,128 +402,148 @@ public static (uint encoded, int bitLength) Encode(int data)
+                } while (i < src.Length && bitsInAcc + 8 <= 64);
 
-                if (j == dst.Length)
+                // Decode bits in accumulator.
+                do
                 {
-                    Array.Resize(ref dstArray, dst.Length * 2);
-                    dst = dstArray;
-                }
+                    lookupIndex = (byte)(acc >> (bitsInAcc - 8));
 
-                dst[j++] = (byte)ch;
+                    ushort lookupValue = decodingTree[(lookupTableIndex << 8) + lookupIndex];
 
-                // If we crossed a byte boundary, advance i so we start at the next byte that's not fully decoded.
-                lastDecodedBits += decodedBits;
-                i += lastDecodedBits / 8;
+                    if (lookupValue < 0x80_00)
+                    {
+                        // Octet found.
+                        // +---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+
+                        // | 0 |     number_of_used_bits   |              octet            |
+                        // +---+---------------------------+-------------------------------+
+                        if (j >= dst.Length)
+                        {
+                            Array.Resize(ref dstArray, dst.Length * 2);
+                            dst = dstArray;
+                        }
+                        dst[j++] = (byte)lookupValue;
 
-                // Modulo 8 since we only care about how many bits were decoded in the last byte that we processed.
-                lastDecodedBits %= 8;
+                        // Start lookup of next symbol
+                        lookupTableIndex = 0;
+                        bitsInAcc -= lookupValue >> 8;
+                    }
+                    else
+                    {
+                        // Traverse to next lookup table.
+                        // +---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+
+                        // | 1 |   next_lookup_table_index |            not_used           |
+                        // +---+---------------------------+-------------------------------+
+                        lookupTableIndex = (lookupValue & 0x7f00) >> 8;
+                        if (lookupTableIndex == 0)
+                        {
+                            // No valid symbol could be decoded or EOS was decoded
+                            throw new HuffmanDecodingException(SR.net_http_hpack_huffman_decode_failed);
+                        }
+                        bitsInAcc -= 8;
+                    }
+                } while (bitsInAcc >= 8);
             }
 
-            return j;
-        }
+            // Finish decoding last < 8 bits of src.
+            // Although bellow code could be inserted in above do..while,
+            // processing of last byte has to handle several corner cases was decided, for performance reasons,
+            // to not have in main decoding loop.
+            while (bitsInAcc > 0)
```

Comment:
Can this go for more than 1 loop?

I'm not questioning the correctness here, just making sure I understand it correctly.

Reply:
Yes it can. Example:
- suppose we last 7 bits left
- 2 bits could form end of currently decoding symbol (1st run)
- 5 bits would be next symbol (2nd run)

I have added unit test test vectors to specifically address this situation.
If one would put break; at then end of this loop, couple of unit tests would fail.

Reply:
> 2 bits could form end of currently decoding symbol (1st run)

This was what I didn't realize could happen. Thanks for explaining.